File: jbake-core/src/main/java/org/jbake/launcher/LaunchOptions.java
package org.jbake.launcher;

import org.jbake.app.configuration.ConfigUtil;
import picocli.CommandLine.ArgGroup;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

import java.io.File;

@Command(
        description = "JBake is a Java based, open source, static site/blog generator for developers & designers",
        name = "jbake",
        usageHelpAutoWidth = true
)
public class LaunchOptions {
    @Parameters(index = "0", description = "source folder of site content (with templates and assets), if not supplied will default to current directory", arity = "0..1")
    private String source;

    @Parameters(index = "1", description = "destination folder for output, if not supplied will default to a folder called \"output\" in the current directory", arity = "0..1")
    private String destination;

    @Option(names = {"-b", "--bake"}, description = "performs a bake")
    private boolean bake;

    @ArgGroup(exclusive = false, heading = "%n%nJBake initialization%n%n")
    private InitOptions initGroup;

    static class InitOptions {

        @Option(names = {"-i", "--init"}, paramLabel = "<target>", description = "initialises required folder structure with default templates (defaults to current directory if <source> is not supplied)", required = true)
        private boolean init;

        @Option(names = {"-t", "--template"}, defaultValue = "freemarker", fallbackValue = "freemarker", description = "use specified template engine for default templates (uses Freemarker if <template> is not supplied) ", arity = "0..1")
        private String template;
    }

    @Option(names = {"-s", "--server"}, description = "runs HTTP server to serve out baked site, if no <value> is supplied will default to a folder called \"output\" in the current directory")
    private boolean runServer;

    @Option(names = {"-c", "--config"}, description = "use specified file for configuration (defaults to " + ConfigUtil.CONFIG_FILE +" in the source folder if not supplied)")
    private String config;

    @Option(names = {"-h", "--help"}, description = "prints this message", usageHelp = true)
    private boolean helpRequested;

    @Option(names = {"--reset"}, description = "clears the local cache, enforcing rendering from scratch")
    private boolean clearCache;

    @Option(names = {"-ls", "--list-settings"}, description = "list configuration settings")
    private boolean listConfig;

    public String getTemplate() {
        return initGroup.template;
    }

    public File getSource() {
        if (source != null) {
            return new File(source);
        } else {
            return new File(System.getProperty("user.dir"));
        }
    }

    public String getSourceValue() {
        return source;
    }

    public File getDestination() {
        if (destination != null) {
            return new File(destination);
        } else {
            return new File(getSource(), "output");
        }
    }

    public String getDestinationValue() {
        return destination;
    }

    public File getConfig() {
        if (config != null) {
            return new File(config);
        } else {
            return new File(getSource(), "jbake.properties");
        }
    }

    public String getConfigValue() {
        return config;
    }

    public boolean isHelpNeeded() {
        return helpRequested || !(isListConfig() || isBake() || isRunServer() || isInit() || source != null || destination != null);
    }

    public boolean isRunServer() {
        return runServer;
    }

    public boolean isInit() {
        return (initGroup !=null && initGroup.init);
    }

    public boolean isClearCache() {
        return clearCache;
    }

    public boolean isBake() {
        return bake || (source != null && destination != null);
    }

    public boolean isListConfig() {
        return listConfig;
    }
}

File: jbake-core/src/main/java/org/jbake/launcher/LaunchOptions.kt
package org.jbake.launcher

import org.jbake.app.configuration.ConfigUtil
import picocli.CommandLine
import java.io.File

@CommandLine.Command(
    description = ["JBake is a Java based, open source, static site/blog generator for developers & designers"],
    name = "jbake",
    usageHelpAutoWidth = true
)
class LaunchOptions {

    @CommandLine.Parameters(index = "0", arity = "0..1",
        description = ["source dir of site content (with templates and assets), if not supplied will default to current directory"],
    )
    var sourceValue: String? = null

    @CommandLine.Parameters(index = "1", arity = "0..1",
        description = ["destination dir for output, if not supplied will default to a dir called \"output\" in the current directory"],
    )
    var destinationValue: String? = null

    @CommandLine.Option(names = ["-b", "--bake"], description = ["performs a bake"])
    var isBake: Boolean = false
        get() = field || (this.sourceValue != null && this.destinationValue != null)

    @CommandLine.ArgGroup(exclusive = false, heading = "%n%nJBake initialization%n%n")
    private var initGroup: InitOptions? = null

    @CommandLine.Option(
        names = ["-s", "--server"],
        description = ["runs HTTP server to serve out baked site, if no <value> is supplied will default to a dir called \"output\" in the current directory"]
    )
    var isRunServer: Boolean = false

    @CommandLine.Option(names = ["-h", "--help"], description = ["prints this message"], usageHelp = true)
    private var helpRequested = false

    @CommandLine.Option(names = ["--reset"], description = ["clears the local cache, enforcing rendering from scratch"])
    var isClearCache: Boolean = false

    @CommandLine.Option(
        names = ["-c", "--config"],
        description = ["use specified file for configuration (defaults to " + ConfigUtil.CONFIG_FILE + " in the source dir if not supplied)"]
    )
    var configValue: String? = null

    @JvmField
    @CommandLine.Option(
        names = ["--prop-encoding"],
        description = ["use given encoding to load properties file. default: utf-8"]
    )
    var propertiesEncoding: String = "utf-8"

    @CommandLine.Option(names = ["-ls", "--list-settings"], description = ["list configuration settings"])
    var isListConfig: Boolean = false

    val template: String?
        get() = initGroup?.template

    fun getSource(): File {
        return sourceValue?.let { File(it) }
            ?: File(System.getProperty("user.dir"))
    }

    fun getDestination(): File {
        return destinationValue?.let { File(it) }
            ?: File(getSource(), "output")
    }

    fun getConfig(): File {
        return configValue?.let { File(it) }
            ?: File(getSource(), "jbake.properties")
    }

    val isHelpNeeded: Boolean
        get() = helpRequested || !(this.isListConfig || this.isBake || this.isRunServer || this.isInit || this.sourceValue != null || this.destinationValue != null)

    val isInit: Boolean
        get() {
            val group = initGroup
            return group != null && group.init
        }

    internal class InitOptions {
        @CommandLine.Option(
            names = ["-i", "--init"],
            paramLabel = "<target>",
            description = ["initialises required dir structure with default templates (defaults to current directory if <source> is not supplied)"],
            required = true
        )
        var init = false

        @CommandLine.Option(
            names = ["-t", "--template"],
            defaultValue = "freemarker",
            fallbackValue = "freemarker",
            description = ["use specified template engine for default templates (uses Freemarker if <template> is not supplied) "],
            arity = "0..1"
        )
        var template: String? = null
    }
}
